File: lab02/prac/lec03/treap.c
#include <assert.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

typedef int64_t data_t;
typedef int64_t prio_t;

typedef struct Node {
    data_t val;
    prio_t priority;
    struct Node *l, *r;
} Node;


prio_t rand_prio() {
    prio_t res = 0;
    for (int i = 0; i < 4; i++) {
        res = (res << 16) ^ rand();
    }
    return res;
}


Node *create_node(data_t val) {
    Node *node = malloc(sizeof(Node));
    node->val = val;
    node->priority = rand_prio();
    node->l = node->r = NULL;
    return node;
}


void split(Node *node, data_t val, Node **l, Node **mid, Node **r) {
    if (node == NULL) {
        *l = *mid = *r = NULL;
    } else if (val == node->val) {
        *l = (node->l);
        *mid = node;
        *r = (node->r);
        node->l = node->r = NULL;
    } else if (val < node->val) {
        *r = node;
        split(node->l, val, l, mid, &(node->l));
    } else {
        assert(val > node->val);
        *l = node;
        split(node->r, val, &(node->r), mid, r);
    }
}


Node *merge(Node *l, Node *r) {
    if (l == NULL) {
        return r;
    }
    if (r == NULL) {
        return l;
    }

    if (l->priority > r->priority) {
        // l becomes the root
        l->r = merge(l->r, r);
        return l;
    } else {
        // r becomes the root
        r->l = merge(l, r->l);
        return r;
    }
}


Node *add(Node *node, data_t val) {
    Node *l, *mid, *r = malloc(sizeof(Node));
    split(node, val, l, mid, r);
    if (!mid) {
        mid = create_node(val);
    }
    return merge(merge(l, mid), r);
}


Node *remove(Node *node, data_t val) {
    Node *l, *mid, *r = malloc(sizeof(Node));
    split(node, val, l, mid, r);
    if (mid) {
        free(mid);
    }
    return merge(l, r);
}


bool contains(Node *node, data_t val) {
    if (!node) {
        return false;
    }

    if (val < node->val) {
        return contains(node->l, val);
    } else if (val > node->val) {
        return contains(node->r, val);
    } else {
        return true;
    }
}


int height(Node *node) {
    if (!node) {
        return 0;
    }
    data_t lh = height(node->l);
    data_t rh = height(node->r);
    return 1 + (lh > rh ? lh: rh);
}


int count_nodes(Node *node) {
    if (!node) {
        return 0;
    }
    return 1 + count_nodes(node->l) + count_nodes(node->r);
}



// class OrderedSet

typedef struct {
    Node *root;
    int size;
} OrderedSet;

void ordered_set_init(OrderedSet *set) {
    set->root = NULL;
}

void ordered_set_add(OrderedSet *set, data_t val) {
    if (!contains(set->root, val)) {
        set->root = add(set->root, val);
        set->size++;
    }
}

void ordered_set_remove(OrderedSet *set, data_t val) {
    if (contains(set->root, val)) {
        set->root = remove(set->root, val);
        set->size--;
    }
}

int ordered_set_count(OrderedSet *set) {
    return count_nodes(set->root);
}

void free_tree(Node *node) {
    if (!node) {
        return;
    }
    free_tree(node->l);
    free_tree(node->r);
    free(node);
}

void ordered_set_free(OrderedSet *set) {
    free_tree(set->root);
    set->root = NULL;
}

int main() {
    int N;
    scanf("%d", &N);

    OrderedSet *set = malloc(sizeof(OrderedSet));
    ordered_set_init(set);

    
    ordered_set_free(set);
    free(set);
    
    return 0;
}

File: lab02/prac/lec03/treap.py
# type: ignore
from dataclasses import dataclass
from random import getrandbits

@dataclass
class Node:
    val: int
    priority: int
    l: "Node | None" = None
    r: "Node | None" = None


def split(node: Node, val: int):
    if node is None:
        return None, None, None
    
    if val == node.val:
        l, r = node.l, node.r
        node.l = node.r = None
        return l, node, r
    elif val < node.val:
        l, mid, node.l = split(node.l, val)
        return l, mid, node
    else:
        assert val > node.val
        node.r, mid, r = split(node.r, val)
        return node, mid, r
    

def merge(l: Node, r: Node):
    if l is None:
        return r
    if r is None:
        return l
    
    if l.priority > r.priority:
        # l will be the root
        l.r = merge(l.r, r)
        return l
    else:
        # r will be the root
        r.l = merge(l, r.l)
        return r
    

def add(node: Node, val: int):
    l, mid, r = split(node, val)
    assert mid is None
    mid = Node(val, priority=getrandbits(64))
    return merge(merge(l, mid), r)


def remove(node: Node, val: int):
    l, mid, r = split(node, val)
    assert mid is None
    # free mid here
    return merge(l, r)


def contains(node: Node, val: int):
    if node is None:
        return False
    
    if val < node.val:
        return contains(node.l, val)
    elif val > node.val:
        return contains(node.r, val)
    else:
        assert val == node.val
        return True


def height(node: Node):
    if node is None:
        return 0
    return 1 + max(height(node.l), height(node.r))


def count_nodes(node: Node):
    if node is None:
        return 0
    return 1 + count_nodes(node.l) + count_nodes(node.r)


class OrderedSet:
    def __init__(self):
        self.root: Node | None = None


    def add(self, val):
        if val not in self:
            self.root = add(self.root, val)


    def remove(self, val):
        if val in self:
            self.root = remove(self.root, val)


    def __contains__(self, val):
        return contains(self.root, val)
    

    def count(self):
        return count_nodes(self.root)
    